### PR TITLE
Course Preview

### LINKED ISSUE #14095: Course assignment: course preview
❌ **This issue is not open for contribution. Visit <a href="https://learningequality.org/contributing-to-our-open-code-base/" target="_blank">Contributing guidelines</a>** to learn about the contributing process and how to find suitable issues.

## Overview

Implement the course preview within the side panel, allowing coaches to view a condensed summary of course content before selecting it for assignment.

**Complexity:** Medium

### Context

| Description | Screenshot |
|---|---|
| A coach can preview a condensed view of the course information within the sidebar | Course preview |

**Existing components to use:**
- `packages/kolibri-common/components/AccordionContainer.vue` and `AccordionItem.vue` - accessibility-tested accordion components
- PR #14077 set up the side panel infrastructure for courses

**Constraints:**
- Coaches cannot directly preview any resources from the assignment side panel
- Coaches cannot directly preview any pre-test or post-test questions

### The Change

Implement the course preview panel showing the course structure:

- Display units with their pre-tests and post-tests (shown automatically as units require them)
- Use accordions to show lessons within each unit
- Display resource count for each lesson
- Fetch course structure using `ContentNodeResource.fetchTree` with the `course_id`

### How to Get There

1. Import a channel with courses (e.g., `gajif-kudos` QA channel)
2. Navigate to Coach > Courses
3. Click to assign a new course
4. Click on a course in the selection list to see the preview

### Out of Scope

- Direct preview of resources from the panel
- Direct preview of pre-test/post-test questions
- Course selection/browsing list (separate issue)
- Recipient selection (separate issue)

### Acceptance Criteria

- [x] Preview displays course units
- [x] Expanding a unit shows its pre-test, lessons, and post-test
- [ ] Pre-test and post-test display their number of questions
- [x] Lessons display their number of resources
- [x] Uses `AccordionContainer`/`AccordionItem` components from kolibri-common
- [x] Shows `KCircularLoader` while course tree data is loading

### References

- [Figma designs](https://www.figma.com/design/MtmNfgSbx8xBlBBJOQmW9g/Pre--post-tests?node-id=2395-121972&t=JBIijX7t0MDS2iFM-0)

---
🤖 This issue was written with AI assistance, under supervision, review and final edits by [@rtibbles] 🤖

### PR DESCRIPTION
**Pull request by nucleogenesis**
## Summary

Adds details for courses

|   |   |   
|---|---|
|image| image | 
| Hidden overflowing long description | image |
| Expanded long description | image |
| w/ thumbnail not expanded | image |
| w/ thumbnail, expanded | image |

## References

Closes #14095

#### QA guidance

- Review the relevant [Figma](https://www.figma.com/design/MtmNfgSbx8xBlBBJOQmW9g/Pre--post-tests?node-id=2452-64030&t=y9ZYEkeyymy7ets1-0)
- Review the requirements of #14095

### CODE REVIEW

**Comment by rtibbles** [comment]
Just from the screenshot - I'm not seeing the pre and post tests here with question counts, and we're using a folder icon instead of the lesson icon for the lessons?

File: kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue [hunk 1]
```diff
@@ -2,12 +2,113 @@
...
+        </div>
+      </section>
+
+      <KCircularLoader v-if="loading" />
+      <AccordionContainer
+        v-else
+        class="course-preview"
+        :headerAppearanceOverrides="{
+          backgroundColor: $themeTokens.surface,
+          fontWeight: 'normal',
```
Comment (rtibbles) [inline kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue:46]: Let's move the non-dynamic styles into the CSS for the course-preview class definition?

File: kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue [hunk 2]
```diff
@@ -2,12 +2,113 @@
...
+          </div>
+        </template>
+        <AccordionItem
+          v-for="(unit, i) in units"
+          :key="unit.id"
+          :title="unit.title"
+          :foldingIconTrailing="false"
+          :headerAppearanceOverrides="{
+            backgroundColor: $themePalette.grey.v_100,
+            fontWeight: 'normal',
```
Comment (rtibbles) [inline kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue]: Likewise here - can add a class for the non-dynamic styles.

File: kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue [hunk 3]
```diff
@@ -2,12 +2,113 @@
...
+            border: `1px solid ${$themeTokens.fineLine}`,
+          }"
+        >
+          <template #content>
+            <ul class='resource-list' :style="{ backgroundColor: $themeTokens.surface }">
+              <li class='resource-item'>
+                <span>
+                  <ContentIcon
+                    :kind="ContentNodeKinds.EXAM"
+                    style="margin-right: 0.5em; font-size: 1.5em;"
```
Comment (rtibbles) [inline kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue]: ibid, and DRY as this is repeated three times!

File: kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue [hunk 4]
```diff
@@ -2,12 +2,113 @@
...
+                <span>{{ numQuestions$({ num: numTestQuestions(unit) }) }}</span>
+              </li>
+              <li
+                v-for="resource in unit.children.results"
+                class='resource-item'
+              >
+                <span>
+                  <ContentIcon
+                    :kind="ContentNodeKinds.LESSON"
+                    style="margin-right: 0.5em; font-size: 1.5em;"
```
Comment (rtibbles) [inline kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue]: ibid, and DRY as this is repeated three times!

File: kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue [hunk 5]
```diff
@@ -2,12 +2,113 @@
...
+                  />
+                  {{ resource.title }}
+                </span>
+                <span>{{ coachStrings.numberOfResources$({ value: resource?.on_device_resources || 0 }) }}</span>
+              </li>
+              <li class='resource-item'>
+                <span>
+                  <ContentIcon
+                    :kind="ContentNodeKinds.EXAM"
+                    style="margin-right: 0.5em; font-size: 1.5em;"
```
Comment (rtibbles) [inline kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue]: ibid, and DRY as this is repeated three times!

File: kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue [hunk 6]
```diff
@@ -29,24 +130,60 @@
...
+      ContentIcon,
       SidePanelLayout,
     },
     setup() {
       const route = useRoute();
       const router = useRouter();
 
       const { backAction$ } = coreStrings;
-      const { courseNameLabel$, selectRecipientsLabel$ } = coursesStrings;
+      const {
```
Comment (rtibbles) [inline kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue:192]: Any reason you've not destructured the coachStrings as well? Seems a little inconsistent to have not done it for those!

**Review by rtibbles — CHANGES_REQUESTED** [review]
I don't think there's anything terribly wrong here per se, but needs a bit of cleanup, mostly related to unnecessary inline styles (which in some cases will cause RTL insensitive behaviour).

Also, I think the heavy use of `em` might prevent us from adhering to our design system guidance for 4px/8px spacing: https://design-system.learningequality.org/layout#spacing

File: kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue (hunk 6 above)
Comment (nucleogenesis) [reply kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue:192]: Nah no good reason - updated it.

File: kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue [hunk 7]
```diff
@@ -64,15 +215,57 @@
         );
       };
 
+      const loading = ref(true);
+      ContentNodeResource.fetchTree({ id: route.params.courseId }).then(results => {
+        course.value = results;
+        loading.value = false;
+      });
```
Comment (rtibbles) [inline kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue]: **blocking:** No error handling for the `fetchTree` call. If the API request fails (e.g., network error, content not available on device), the component will stay in a permanent loading state since `loading` is never set to `false` on failure.

Consider adding a `.catch()` handler:
```js
ContentNodeResource.fetchTree({ id: route.params.courseId })
  .then(results => {
    course.value = results;
    loading.value = false;
  })
  .catch(() => {
    loading.value = false;
    // Consider setting an error state to show a message
  });
```

File: kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue [hunk 8]
```diff
@@ -64,15 +215,57 @@
...
+
+      const courseSubtitle = computed(() => {
+        if (loading.value == true) {
```
Comment (rtibbles) [inline kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue]: **suggestion:** 
```js
if (loading.value) {
```

File: kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue [hunk 9]
```diff
@@ -64,15 +215,57 @@
...
+        numberOfResources$,
+        viewMoreAction$,
+        viewLessAction$,
+        ContentNodeKinds,
+
+        course,
+        units,
       };
     },
+    created() {},
```
Comment (rtibbles) [inline kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue]: **suggestion:** Empty `created()` lifecycle hook should be removed.

File: kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue [hunk 10]
```diff
@@ -2,12 +2,129 @@
...
+                  <ContentIcon
+                    class="content-icon"
+                    :kind="ContentNodeKinds.EXAM"
+                  />
+                  {{ preTestLabel$() }}
+                </span>
+                <span>{{ numQuestions$({ num: numTestQuestions(unit) }) }}</span>
+              </li>
+              <li
+                v-for="resource in unit.children.results"
```
Comment (rtibbles) [inline kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue]: **suggestion:** `unit.children.results` is accessed in multiple places (lines 92, 103, 123) without defensive checks. If the API returns a unit without `children` or `children.results`, this would throw. Consider a helper or guard:
```js
const getUnitLessons = (unit) => unit?.children?.results || [];
```

File: packages/kolibri-common/strings/coursesStrings.js [hunk 11]
```diff
@@ -122,4 +122,36 @@ export const coursesStrings = createTranslator('CoursesStrings', {
...
+    message: '{num, number} {num, plural, one {lesson} other {lessons}}',
+    context: 'Part of course details on the heading for each unit showing how much is inside',
+  },
+  numUnits: {
+    message: '{num, number} {num, plural, one {unit} other {units}}',
+    context: 'Part of subtitle shown under the course title',
+  },
+  numResources: {
+    message: '{num, number} {num, plural, one {resource} other {resources}}',
+    context: 'Part of subtitle shown under the course title',
```
Comment (rtibbles) [inline packages/kolibri-common/strings/coursesStrings.js]: **blocking:** The `numResources` string is added here but never used in this PR. The subtitle uses `numberOfResources$` from `coachStrings` instead (CourseDetails.vue line 230). This dead string will create unnecessary translation work.

Either:
1. Remove `numResources` from here, or
2. Use it in the component instead of `numberOfResources$` from coachStrings (which would consolidate all course-related strings in one place)

File: packages/kolibri-common/components/accordion/AccordionItem.vue [hunk 12]
```diff
@@ -20,6 +20,13 @@
       >
         <div class="header-content">
           <div class="title-actions-wrapper">
+            <KIconButton
+              v-if="!foldingIconTrailing"
+              tabindex="-1"
+              :icon="isExpanded ? 'chevronDown' : 'chevronRight'"
+              style="margin-right: 0.5em"
+              @click.stop="toggle"
```
Comment (rtibbles) [inline packages/kolibri-common/components/accordion/AccordionItem.vue]: **suggestion:** The leading `KIconButton` has `@click.stop="toggle"` but it sits inside the header `<button>` which also has `@click.stop="toggle"`. While `.stop` prevents the icon click from bubbling to the parent button, the icon click handler still calls `toggle` directly, so it works correctly. However, since the parent button already handles toggle, you could simplify the icon to be purely visual by removing its `@click.stop` handler. This would also avoid any edge case where both handlers could fire.

**Review by rtibbles — COMMENTED** [review]
Nice work implementing the course preview panel - the accordion-based unit display with pre/post test details looks well-structured and matches the design intent. The description expand/collapse and the leading accordion icon are nice touches. A few items below to address before merge.

**Blocking:**
- No error handling for failed `fetchTree` API call
- Unused `numResources` string added to `coursesStrings`

**Suggestions/Nitpicks:**
- Check boolean value directly
- Defensive access on `unit.children.results`
- Remove empty `created()` hook
- Accordion icon click handler overlap

Looking forward to the updates - let me know if you have questions on any of these!

This review was conducted by Claude Code with final review by @rtibbles.

File: kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue [hunk 13]
```diff
@@ -2,12 +2,131 @@
...
+            v-if="course.thumbnail"
+            :src="course.thumbnail"
+            isDecorative
+          />
+        </div>
+        <div
+          ref="courseDescriptionRef"
+          class="course-description"
+          :style="{
+            paddingLeft: course.thumbnail ? '16px' : '0',
```
Comment (rtibbles) [inline kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue]:
```
['padding' + isRtl ? 'Right' : 'Left']: course.thumbnail ? '16px' : '0',
```
Needs to be RTL sensitive.

File: kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue [hunk 14]
```diff
@@ -2,12 +2,131 @@
...
+            maxHeight: descExpanded ? 'none' : '136px',
+          }"
+        >
+          {{ course.description }}
+        </div>
+      </section>
+      <KButton
+        v-if="descOverflowing"
+        style="display: block; margin-bottom: 32px; text-align: right"
```
Comment (rtibbles) [inline kolibri/plugins/coach/frontend/views/courses/sidePanels/AssignCourse/subpages/CourseDetails.vue]: Let's turn this into a class and put in CSS, so that it will get properly handled for RTL.

File: packages/kolibri-common/components/accordion/AccordionItem.vue [hunk 15]
```diff
@@ -20,6 +20,12 @@
       >
         <div class="header-content">
           <div class="title-actions-wrapper">
+            <KIconButton
+              v-if="!foldingIconTrailing"
+              tabindex="-1"
+              :icon="isExpanded ? 'chevronDown' : 'chevronRight'"
+              style="margin-right: 0.5em"
```
Comment (rtibbles) [inline packages/kolibri-common/components/accordion/AccordionItem.vue]: Move into CSS for RTL, change to `px` for non-relative usage, as per KDS standards.

**Review by rtibbles — CHANGES_REQUESTED** [review]
Still a few style cleanups needed here.

There was one oversight in the spec - when the accordions are all expanded, we should probably switch the button to 'Collapse all units'!

Image

**Comment by nucleogenesis** [comment]
@rtibbles I've updated the expand/collapse all as discussed

https://github.com/user-attachments/assets/762249ec-472f-4732-b442-f16bf71791db

**Review by rtibbles — APPROVED** [review]
All comments addressed, quick manual QA checks out.
